File: apps/bridge/src/NfeAgendamento.Bridge/Fiscal/AccessKey.cs
namespace NfeAgendamento.Bridge.Fiscal;

public sealed record AccessKey
{
    private AccessKey(string value)
    {
        Value = value;
        UfAutor = value[..2];
    }

    public string Value { get; }

    public string UfAutor { get; }

    public static bool TryParse(string? value, out AccessKey? accessKey)
    {
        accessKey = null;
        if (string.IsNullOrWhiteSpace(value)) return false;

        var normalized = value.Trim().ToUpperInvariant();
        if (normalized.Length != 44) return false;

        for (var index = 0; index < normalized.Length; index++)
        {
            var character = normalized[index];
            var alphaNumericCnpjPosition = index is >= 6 and < 18;

            if (alphaNumericCnpjPosition)
            {
                if (!char.IsAsciiDigit(character) && character is not (>= 'A' and <= 'Z')) return false;
            }
            else if (!char.IsAsciiDigit(character))
            {
                return false;
            }
        }

        var sum = 0;
        var weight = 2;
        for (var index = 42; index >= 0; index--)
        {
            // NT Conjunta DFe 2025.001: caracteres alfanuméricos usam valor ASCII - 48.
            sum += (normalized[index] - '0') * weight;
            weight = weight == 9 ? 2 : weight + 1;
        }

        var checkDigit = 11 - (sum % 11);
        if (checkDigit >= 10)
        {
            checkDigit = 0;
        }

        if (normalized[43] - '0' != checkDigit)
        {
            return false;
        }

        accessKey = new AccessKey(normalized);
        return true;
    }

    public override string ToString() => Value;
}

File: apps/web/src/nfe/access-key.ts
export type AccessKeyValidation =
  | { valid: true; value: string; ufAutor: string }
  | { valid: false; error: string };

const ACCESS_KEY_PATTERN = /^[0-9]{6}[A-Z0-9]{12}[0-9]{26}$/;
const NFE_MODEL = '55';

export function validateAccessKey(value: string): AccessKeyValidation {
  const normalized = value.trim().toUpperCase();

  if (normalized.length !== 44) {
    return { valid: false, error: 'A chave da NF-e deve conter 44 caracteres.' };
  }

  if (!ACCESS_KEY_PATTERN.test(normalized)) {
    return {
      valid: false,
      error: 'A chave da NF-e não está no formato oficial (A-Z e 0-9).',
    };
  }

  if (normalized.slice(20, 22) !== NFE_MODEL) {
    return { valid: false, error: 'Este sistema aceita somente NF-e modelo 55.' };
  }

  const checkDigit = calculateCheckDigit(normalized.slice(0, 43));
  if (Number(normalized[43]) !== checkDigit) {
    return { valid: false, error: 'Dígito verificador da chave NF-e inválido.' };
  }

  return {
    valid: true,
    value: normalized,
    ufAutor: normalized.slice(0, 2),
  };
}

function calculateCheckDigit(prefix: string): number {
  let sum = 0;
  let weight = 2;

  for (let index = prefix.length - 1; index >= 0; index -= 1) {
    // NT Conjunta DFe 2025.001: caracteres alfanuméricos usam valor ASCII - 48.
    sum += (prefix.charCodeAt(index) - 48) * weight;
    weight = weight === 9 ? 2 : weight + 1;
  }

  const digit = 11 - (sum % 11);
  return digit >= 10 ? 0 : digit;
}
